contrib/examples/actions/workflows/customfilters/mistral-regex_match.yaml

```diff
@@ -0,0 +1,20 @@
+version: '2.0'
+
+examples.mistral-customfilters-regex_match:
+    description: A workflow for testing regex_match custom filter in mistral
+    type: direct
+    vars:
+        regex_pattern: "([0-9]{3} \\w+ (?:Ave|St|Dr))"
+    input:
+        - input_str
+    output:
+        result_jinja: <% $.result_jinja %>
+        result_yaql: <% $.result_yaql %>
+    tasks:
+
+        task1:
+            action: std.noop
+            publish:
+                result_jinja: '{{ _.input_str | regex_match(_.regex_pattern)}}'
```

Comment:
Sure thing. So, I threw together a quick filter to simply print out the positional args that are being passed into it:

```
def test_filter(*args):
    return str(args)
```

Then put together a workflow for testing (ignore the regex naming, I just used that as a testbed):

```
version: '2.0'

examples.mistral-customfilters-regex_match:
    description: A workflow for testing regex_match custom filter in mistral
    type: direct
    input:
        - input_str
        - regex_pattern
    output:
        result_jinja: <% $.result_jinja %>
        result_yaql: <% $.result_yaql %>
    tasks:

        task1:
            action: std.noop
            publish:
                result_jinja: '{{ _.input_str | test_filter }}'
                result_yaql: '<% test_filter($.input_str) %>'
```

When we use the pipe (in Jinja's eyes, filter usage), the first arg is simply the string value being passed in:

```
(virtualenv)vagrant@st2dev:~/st2$ segl
id: 597abe9232ed355b4e873d3e
action.ref: examples.mistral-customfilters-regex_match
parameters:
  input_str: Your address is 567 Elsewhere Dr. My address is 123 Somewhere Ave.
  regex_pattern: ([0-9]{3} \w+ (?:Ave|St|Dr))
status: succeeded (5s elapsed)
start_timestamp: 2017-07-28T04:33:22.578088Z
end_timestamp: 2017-07-28T04:33:27.492128Z
result:
  extra:
    state: SUCCESS
    state_info: null
  result_jinja: (u'Your address is 567 Elsewhere Dr. My address is 123 Somewhere Ave.',)
  result_yaql: (u'Your address is 567 Elsewhere Dr. My address is 123 Somewhere Ave.',)
  tasks:
  - created_at: '2017-07-28 04:33:23'
    id: 213d3c60-b380-49e3-a725-6b76f6f84c0c
    input: null
    name: task1
    published:
      result_jinja: (u'Your address is 567 Elsewhere Dr. My address is 123 Somewhere Ave.',)
      result_yaql: (u'Your address is 567 Elsewhere Dr. My address is 123 Somewhere Ave.',)
    result: null
    state: SUCCESS
    state_info: null
    updated_at: '2017-07-28 04:33:23'
    workflow_execution_id: 37d5f7b6-5af2-445a-b5fe-d72cd490c1c3
    workflow_name: examples.mistral-customfilters-regex_match
```

However, if we change the Jinja snippet to behave like a function instead of a filter....

```
'{{ test_filter(_.input_str) }}'
```

...then the first parameter will be the context, followed by the string value:

```
(virtualenv)vagrant@st2dev:~/st2$ segl
id: 597abfc332ed356072d3b2f1
action.ref: examples.mistral-customfilters-regex_match
parameters:
  input_str: Your address is 567 Elsewhere Dr. My address is 123 Somewhere Ave.
  regex_pattern: ([0-9]{3} \w+ (?:Ave|St|Dr))
status: succeeded (5s elapsed)
start_timestamp: 2017-07-28T04:38:27.931985Z
end_timestamp: 2017-07-28T04:38:32.919549Z
result:
  extra:
    state: SUCCESS
    state_info: null
  result_jinja: ({}, u'Your address is 567 Elsewhere Dr. My address is 123 Somewhere Ave.')
  result_yaql: (u'Your address is 567 Elsewhere Dr. My address is 123 Somewhere Ave.',)
  tasks:
  - created_at: '2017-07-28 04:38:28'
    id: bd3568b5-ed16-494e-8ca2-c76cb76227db
    input: null
    name: task1
    published:
      result_jinja: ({}, u'Your address is 567 Elsewhere Dr. My address is 123 Somewhere Ave.')
      result_yaql: (u'Your address is 567 Elsewhere Dr. My address is 123 Somewhere Ave.',)
    result: null
    state: SUCCESS
    state_info: null
    updated_at: '2017-07-28 04:38:28'
    workflow_execution_id: 6dc96bc0-f8e0-47d5-bf74-60c313715e12
    workflow_name: examples.mistral-customfilters-regex_match
```

The reason this is happening is that when used as a function, Jinja2 injects it's own Context object as the very first parameter. What complicates things is that Mistral is ALSO passing in its own context parameter (the empty dict we see in the previous example) all the time, but the use of `partial` in https://github.com/StackStorm/mistral/blob/master/mistral/utils/expression_utils.py#L92 means we usually don't see this. However, if we use this as a function, the Jinja2 Context object becomes args[0], and Mistral's context becomes args[1], leaving value at args[2].

So, TLDR the reason the examples (see the ActionChains in examples pack, for instance - their back-end functions are implemented in the same way) use the pipe notation is because all of the backend functions have the string value as the first parameter. It would seem that the current implementation requires the pipe notation.

Now, this obviously doesn't HAVE to be the case - I could, for instance, create a decorator to handle these variations seamlessly (and/or use one of the decorators at http://jinja.pocoo.org/docs/2.9/api/#utilities), but whatever we do for Jinja in Mistral should get copied over to st2 so we don't undo the work to sync the two "islands".

Hopefully that makes sense. I'll take a stab at trying to make the underlying functions care less about how they're called, keeping in mind that it will have to also be repeated in st2.